File: dsa/cpp/BST.cpp
#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* searchBST(TreeNode* root, int val) {
        if (root == nullptr) {
            return nullptr;
        }
        if (root->val == val) {
            return root;
        }

        if (root->val < val) {
            return searchBST(root->right, val);  // RETURN this
        } else {
            return searchBST(root->left, val);   // RETURN this
        }
    }
};


void printSubtree(TreeNode* node) {
    if (!node) return;
    cout << node->val << " ";
    printSubtree(node->left);
    printSubtree(node->right);
}

int main() {
    // Construct this tree:
    //        5
    //       / \
    //      3   7
    //     / \   \
    //    2   4   8

    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(2);
    root->left->right = new TreeNode(4);
    root->right->right = new TreeNode(8);

    Solution solver;
    int val = 7; // Change this to test other values
    TreeNode* result = solver.searchBST(root, val);

    if (result) {
        cout << "Subtree with root " << val << ": ";
        printSubtree(result); // prints the subtree starting from that node
    } else {
        cout << "Value not found in BST.";
    }

    return 0;
}

File: dsa/python/BST.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def searchBST(self, root, val):
        if root is None or root.val == val:
            return root
        if root.val > val:
            return self.searchBST(root.left, val)
        else:
            return self.searchBST(root.right, val)

def insert_node(root, val):
    if root is None:
        return TreeNode(val)
    if val < root.val:
        root.left = insert_node(root.left, val)
    else:
        root.right = insert_node(root.right, val)
    return root

def build_bst(values):
    root = None
    for val in values:
        root = insert_node(root, val)
    return root

if __name__ == "__main__":
    values = [10, 5, 15, 3, 7, 12, 18]
    root = build_bst(values)
    val_to_search = 7
    sol = Solution()
    result = sol.searchBST(root, val_to_search)
    if result:
        print("Node found with value:", result.val)
    else:
        print("Value not found")
